### PR TITLE
Add joining and leaving projects

### PR DESCRIPTION
**Pull request by Melissa0x1f992**
Implements /projects/[uuid]/membership from the hackmd

### CODE REVIEW

File: pages/projects.js [hunk 1]
```diff
@@ -20,18 +21,31 @@ const ProjectCard = styled.div`
...
 const ProjectPreview = ({ project }) => {
+	const toggleProjectMembership = async (event) => {
+		event.preventDefault();
+
+		makeApiRequest(`/y22/projects/${project.uuid}/membership`, {
+			method: project.joined ? "DELETE" : "PUT",
+		});
+
+		// This is gross and pain.
+		Router.reload(window.location.pathname);
```
Comment (haykam821) [inline pages/projects.js:32]: Could this modify the project state instead?

**Review by haykam821 — DISMISSED** [review]

File: lib/api.js [hunk 2]
```diff
@@ -1,7 +1,8 @@
 export const API_BASE = "https://api.snakeroom.org";
 
-export const makeApiRequest = (path) => {
+export const makeApiRequest = (path, fetchParams) => {
 	return fetch(API_BASE + path, {
-		credentials: "include",
+		...{ credentials: "include" },
```
Comment (opl-) [inline lib/api.js:5]: Why not just this?
```suggestion
		credentials: "include",
```

File: pages/projects.js [hunk 3]
```diff
@@ -19,19 +20,81 @@ const ProjectCard = styled.div`
...
-					disabled
-				/>
+			{typeof membershipStatus === "object" && (
+				<form onSubmit={membershipStatus === MembershipStatus.Joined ? leaveProject : joinProject}>
+					<SubmitButton
+						type="submit"
+						value={
+							membershipStatus === MembershipStatus.Joined ?
+							"Leave" : membershipStatus === MembershipStatus.NotJoined ?
+							"Join" : "Error"
```
Comment (opl-) [inline pages/projects.js:94]: Weird formatting. I'd expect the string to be on the same line as the condition.

File: pages/projects.js [hunk 4]
```diff
@@ -19,19 +20,81 @@ const ProjectCard = styled.div`
...
+	static Unknown = new MembershipStatus("Unknown");
+
+	constructor(text) {
+		this.text = text;
+	}
+}
+
 const ProjectPreview = ({ project }) => {
+	let initialMembershipStatus = undefined;
+	if (typeof project.joined === "boolean") {
```
Comment (opl-) [inline pages/projects.js:37]: So if this isn't returned by the API, the user just can't join nor leave?

**Review by opl- — CHANGES_REQUESTED** [review]

**Pull request merged**